runtime/commonMain/src/kotlinx/serialization/json/JsonTransformingSerializer.kt

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2017-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.serialization.json
+
+import kotlinx.serialization.*
+import kotlinx.serialization.internal.*
+import kotlinx.serialization.json.internal.*
+
+/**
+ * Base class for custom serializers that allows manipulating an abstract JSON
+ * representation of the class.
+ *
+ * Instead of direct interaction with [Encoder] or [Decoder], this class
+ * offers you to manipulate [JsonElement] instead, what can be much easier
+ * way to accomplish certain tasks. Please note that this class expects that
+ * [Encoder] and [Decoder] are implemented by [JsonInput] and [JsonOutput], i.e.
+ * serializers derived from this class work only with [Json] format.
+ *
+ * During serialization, this class first serializes original value with [tSerializer] to a [JsonElement],
+ * then calls [writeTransform] method, which may contain a user-defined transformation, such as
+ * wrapping a value into [JsonArray], filtering keys, adding keys, etc.
+ *
+ * During deserialization, reverse process happens: first, value from JSON stream is read
+ * to a [JsonElement], then user transformation in [readTransform] applied,
+ * then JSON tree is deserialized back to [T] with [tSerializer].
+ *
+ * Usage example:
+ *
+ * ```
+ * @Serializable
+ * data class Example(
+ *     @Serializable(UnwrappingJsonListSerializer::class) val data: String
+ * )
+ * // Unwraps a list to a single object
+ * object UnwrappingJsonListSerializer :
+ *     JsonTransformingSerializer<String>(String.serializer(), "UnwrappingList") {
+ *     override fun readTransform(element: JsonElement): JsonElement {
+ *         if (element !is JsonArray) return element
+ *         require(element.size == 1) { "Array size must be equal to 1 to unwrap it" }
+ *         return element.first()
+ *     }
+ * }
+ * // Now these functions both yield correct result:
+ * Json.parse(Example.serializer(), """{"data":["str1"]}""")
+ * Json.parse(Example.serializer(), """{"data":"str1"}""")
+ * ```
+ *
+ * @param T A type for Kotlin property for which this serializer could be applied.
+ *        **Not** the type that you may encounter in JSON. (e.g. if you unwrap a list
+ *        to a single value `T`,use `T`, not `List<T>`)
+ * @param tSerializer A serializer for type [T]. Determines [JsonElement] which is passed to [writeTransform].
+ *        Should be able to parse [JsonElement] from [readTransform] function.
+ *        Usually, default serializer is sufficient.
+ * @param transformationName A name for the particular implementation to fulfill [SerialDescriptor.serialName] uniqueness guarantee.
+ */
+abstract class JsonTransformingSerializer<T : Any>(
+    private val tSerializer: KSerializer<T>,
+    transformationName: String
+) : KSerializer<T> {
```

Comment:
The class itself is great and I'd expected it to be used frequently.
The only concern is a not-so-short name. WDYT about "JsonTransformer" instead?

Reply:
We have 'Serializer' word appended to all `KSerializer`s, AFAIR. I do not have any objections to `JsonTransformer` besides that. 
Also, this name occurs only in supertype position, so maybe we shouldn't pursue shortness at all

Reply:
I am fine with both options